src/helper.cc

```diff
@@ -848,20 +870,27 @@ helper::handleKilledServer(HelperServerBase *srv, bool &needsNewServers)
         debugs(84, DBG_CRITICAL, "WARNING: " << id_name << " #" << srv->index << " exited");
 
         if (childs.needNew() > 0) {
-            debugs(80, DBG_IMPORTANT, "Too few " << id_name << " processes are running (need " << childs.needNew() << "/" << childs.n_max << ")");
-
-            if (childs.n_active < childs.n_startup && last_restart > squid_curtime - 30) {
-                if (srv->stats.replies < 1)
-                    fatalf("The %s helpers are crashing too rapidly, need help!\n", id_name);
-                else
-                    debugs(80, DBG_CRITICAL, "ERROR: The " << id_name << " helpers are crashing too rapidly, need help!");
-            }
+            handleLackOfServers(srv->stats.replies >= 1);
             srv->flags.shutdown = true;
             needsNewServers = true;
         }
     }
 }
 
+void
+helper::handleLackOfServers(const bool madeProgress)
+{
+    debugs(80, DBG_IMPORTANT, "Too few " << id_name << " processes are running (need " << childs.needNew() << "/" << childs.n_max << ")" <<
+           Debug::Extra << "active processes: " << childs.n_active <<
+           Debug::Extra << "processes configured to start at (re)configuration: " << childs.n_startup);
+    if (childs.n_active < childs.n_startup && last_restart > squid_curtime - 30) {
```

Comment:
Just to separate "too few are running" problem handling from "crashing too rapidly" problem handling:

```suggestion

    if (childs.n_active < childs.n_startup && last_restart > squid_curtime - 30) {
```

The official code did have that separation.

Reply:
Ok.